### PR TITLE
Type-annotate `wandb.integration.tensorboard`

### LINKED ISSUE #3229: [CLI] [App] Multiple tensorboard issues.
**Description**
The tensorboard integration has a number of issues.

## Issue 1: `global_step` reset.

Related to #1997, when running the following code with the latest `wandb=0.12.10`, nothing gets logged as shown in this [tracked experiment](https://wandb.ai/costa-huang/cleanRL/runs/17g4fl38)

```python
import numpy as np
import random
import time
from torch.utils.tensorboard import SummaryWriter
import wandb

if __name__ == "__main__":
    # setup the environment
    experiment_name = f"wandb=={wandb.__version__}__{int(time.time())}"
    run = wandb.init(
        project="cleanRL",
        entity=None,
        sync_tensorboard=True,
        name=experiment_name,
        save_code=True,
    )
    writer = SummaryWriter(f"runs/{experiment_name}")
    for global_step in range(100):
        writer.add_scalar("loss1", np.random.rand(1)[0], global_step)
    for global_step in range(100):
        writer.add_scalar("loss2", np.random.rand(1)[0], global_step)
```

image

Doing it slightly differently using `wandb.tensorboard.patch(save=False)` also results in [undesired results](https://wandb.ai/costa-huang/cleanRL/runs/39afcf46), where `loss2` is displayed incorrectly 

```diff
 import numpy as np
 import random
 import time
 from torch.utils.tensorboard import SummaryWriter
 import wandb
 
 if __name__ == "__main__":
     # setup the environment
     experiment_name = f"wandb=={wandb.__version__}__{int(time.time())}"
     run = wandb.init(
         project="cleanRL",
         entity=None,
-        sync_tensorboard=True,
         name=experiment_name,
         save_code=True,
     )
+    wandb.tensorboard.patch(save=False)
     writer = SummaryWriter(f"runs/{experiment_name}")
     for global_step in range(100):
         writer.add_scalar("loss1", np.random.rand(1)[0], global_step)
     for global_step in range(100):
         writer.add_scalar("loss2", np.random.rand(1)[0], global_step)
```

image

#1997's suggestion fo adding `wandb.tensorboard.patch(root_logdir=f'runs/{experiment_name}')` which fixes the issue as shown in this [tracked experiment](https://wandb.ai/costa-huang/cleanRL/runs/2skd9qyi), but it feels a bit hacky... 

```diff
 import numpy as np
 import random
 import time
 from torch.utils.tensorboard import SummaryWriter
 import wandb
 
 if __name__ == "__main__":
     # setup the environment
     experiment_name = f"wandb=={wandb.__version__}__{int(time.time())}"
+    wandb.tensorboard.patch(root_logdir=f'runs/{experiment_name}') 
     run = wandb.init(
         project="cleanRL",
         entity=None,
-        sync_tensorboard=True,
         name=experiment_name,
         save_code=True,
     )
     writer = SummaryWriter(f"runs/{experiment_name}")
     for global_step in range(100):
         writer.add_scalar("loss1", np.random.rand(1)[0], global_step)
     for global_step in range(100):
         writer.add_scalar("loss2", np.random.rand(1)[0], global_step)
```
image

## Issue 2: missing data when displayed using `global_step` as x-axis

The successful experiment that incorporated the approach from #1997 still has a minor issue: in the dashboard, we don't see the datapoint in `loss1` when `x=99` (we can only see up to `x=98`). See the video below for a demo.

https://user-images.githubusercontent.com/5555347/152885010-c669da4f-28e0-4eed-8db7-2fac65a3117a.mp4

While this may appear like a minor issue, I run into a similar issue like this at a much larger scale. As shown in the video below, in [this experiment](https://wandb.ai/gym-microrts/gym-microrts/runs/13tizcke?workspace=user-costa-huang) I was missing data that are available when using `Step` as the x-axis. Clearly, the last `charts/trueskill` summary is 37.323, but I could not visualize this data even when zooming into the chart that used the `global_step` as the x-axis.

https://user-images.githubusercontent.com/5555347/152885830-c2b39863-01eb-4522-a7df-cc0e799ab212.mp4

## Issue 3: data not logged in Windows with `sync_tensorboard=True`

A number of my collaborators who use Windows have tried to log experiments with `wandb.init(..., sync_tensorboard=True)` but no data was logged, which looks like

image

See https://github.com/vwxyzjn/gym-microrts/pull/25 and https://github.com/vwxyzjn/vectorized-value-methods/pull/7. What worked for them on Windows is to replace `sync_tensorboard=True` with `wandb.tensorboard.patch(save=False)`. 

## Issue 4: Tensorboard page in the app crashes:

https://wandb.ai/costa-huang/cleanRL/runs/2skd9qyi/tensorboard?workspace=user-costa-huang

See the following video

https://user-images.githubusercontent.com/5555347/152886944-326807b2-9336-4cbd-81f4-05066812fde0.mp4

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------
This PR was started to explore the apparent issues reported in https://github.com/wandb/client/issues/3229 and the corresponding Jira tickets, but since those look like something outside of our control, turned into just type-annotating `wandb.integration.tensorboard`.

Testing
-------
Green CI!

### CODE REVIEW

File: wandb/integration/tensorboard/__init__.py [hunk 1]
```diff
@@ -2,7 +2,14 @@
...
 
+from .log import _log, log, reset_state, tf_summary_to_dict
 from .monkeypatch import patch, unpatch
-from .log import _log, log, tf_summary_to_dict, reset_state
 
-__all__ = ["patch"]
+__all__ = [
+    "patch",
+    "unpatch",
+    "_log",
```
Comment (kptkin) [inline wandb/integration/tensorboard/__init__.py:11]: do we want to include internal function?

File: wandb/integration/tensorboard/__init__.py [hunk 2]
```diff
@@ -2,7 +2,14 @@
...
+    "log",
+    "reset_state",
+    "tf_summary_to_dict",
```
Comment (kptkin) [inline wandb/integration/tensorboard/__init__.py:14]: this one as well... I don't think we want to import it by defualt?

File: wandb/integration/tensorboard/__init__.py [hunk 3]
```diff
@@ -2,7 +2,14 @@
...
 from .monkeypatch import patch, unpatch
-from .log import _log, log, tf_summary_to_dict, reset_state
 
-__all__ = ["patch"]
+__all__ = [
+    "patch",
+    "unpatch",
+    "_log",
+    "log",
+    "reset_state",
```
Comment (kptkin) [inline wandb/integration/tensorboard/__init__.py:13]: not sure what this one is doing...

File: wandb/integration/tensorboard/log.py [hunk 4]
```diff
@@ -320,22 +330,24 @@ def _log(
             or timestamp - STEPS["global"]["last_log"] >= RATE_LIMIT_SECONDS
         ):
             if history is None:
-                wandb.run._log({})
+                if wandb.run is not None:
```
Comment (kptkin) [inline wandb/integration/tensorboard/log.py:333]: probably a better place to check this in the estimator_hook: https://github.com/wandb/client/blob/master/wandb%2Fintegration%2Ftensorflow%2Festimator_hook.py#L48

we can add a warning message that they need to init the run before calling the hook....
like all of our other integrations

File: wandb/integration/tensorboard/log.py [hunk 5]
```diff
@@ -320,22 +330,24 @@ def _log(
...
+                    wandb.run._log({})
             else:
-                history.add({}, **kwargs)
+                history.add({})
 
         STEPS["global"]["last_log"] = timestamp
 
     if history is None:
-        wandb.run._log(log_dict, commit=False)
+        if wandb.run is not None:
```
Comment (kptkin) [inline wandb/integration/tensorboard/log.py:341]: see previous comment, maybe in this function we could just raise an assertion error (if history is `None`)?

File: wandb/integration/tensorboard/monkeypatch.py [hunk 6]
```diff
@@ -122,23 +129,27 @@ def new_csfw_func(*args, **kwargs):
     wandb.patched["tensorboard"].append([module, "create_summary_file_writer"])
 
 
-def _patch_file_writer(writer, module, save=None, root_logdir=None):
+def _patch_file_writer(
+    writer: Any, module: Any, save: bool = True, root_logdir: str = "",
+) -> None:
     # This configures non-TensorFlow Tensorboard logging, or tensorflow <= 1.15
     old_efw_class = writer.EventFileWriter
```
Comment (kptkin) [inline wandb/integration/tensorboard/monkeypatch.py:136]: since this is a class, maybe we should rename it to match python conventions for naming classes?

File: wandb/integration/tensorboard/monkeypatch.py [hunk 7]
```diff
@@ -163,5 +173,8 @@ def __init__(self, logdir, *args, **kwargs):
     wandb.patched["tensorboard"].append([module, "EventFileWriter"])
 
 
-def _notify_tensorboard_logdir(logdir, save=None, root_logdir=None):
-    wandb.run._tensorboard_callback(logdir, save=save, root_logdir=root_logdir)
+def _notify_tensorboard_logdir(
+    logdir: str, save: bool = True, root_logdir: str = ""
+) -> None:
+    if wandb.run is not None:
```
Comment (kptkin) [inline wandb/integration/tensorboard/monkeypatch.py:179]: happens if `wandb.run` is `None` should we just continue?

**Comment by kptkin** [comment]
oh, also since you already have the examples from the ticket can we convert it to functional tests?

File: wandb/integration/tensorboard/log.py (hunk 4 above)
Comment (dmitryduev) [reply wandb/integration/tensorboard/log.py:333]: Changed to raising an error, like everywhere else. 
This should really never happen, it's mainly for mypy since `wandb.run` is `Optional`.

File: wandb/integration/tensorboard/log.py (hunk 5 above)
Comment (dmitryduev) [reply wandb/integration/tensorboard/log.py:341]: same as above

File: wandb/integration/tensorboard/monkeypatch.py (hunk 7 above)
Comment (dmitryduev) [reply wandb/integration/tensorboard/monkeypatch.py:179]: I thought that it only happens during `wandb.init`, but then remembered that I've seen code where people call it before init'ing, perhaps we've been implicitly relying on late binding (the relevant call lives in this `TBXEventFileWriter` class) Let me try and raise an error like everywhere else, let's see what the CI has to say about it...

File: wandb/integration/tensorboard/log.py [hunk 8]
```diff
@@ -320,13 +330,21 @@ def _log(
             or timestamp - STEPS["global"]["last_log"] >= RATE_LIMIT_SECONDS
         ):
             if history is None:
+                if wandb.run is None:
+                    raise wandb.Error(
+                        "You must call `wandb.init()` before calling `wandb.tensorflow._log`"
```
Comment (kptkin) [inline wandb/integration/tensorboard/log.py:335]: hmm... i don't think this right to do it here... since it is an internal function that the user doesn't suppose to call... we call it under the hood [here](https://github.com/wandb/client/blob/master/wandb/integration/tensorboard/log.py#L344) and [here](https://github.com/wandb/client/blob/master/wandb/integration/tensorflow/estimator_hook.py#L48) (relevant documentation [estimator-hook](https://docs.wandb.ai/guides/integrations/tensorflow#tensorflow-hook) and [tf-log](https://docs.wandb.ai/guides/integrations/tensorflow#manual-logging)).
We have this message for `tensorflow.log` maybe we can add it in the `estimator hook` constructor, what do you think? here we could have an assertion in that basically checks if history is `None` wandb.run can't be `None`?

Comment (kptkin) [reply wandb/integration/tensorboard/log.py:335]: for the estimator hook: looking at the [begin](https://www.tensorflow.org/api_docs/python/tf/estimator/SessionRunHook#begin) function of the `SessionRunHook`... maybe this is the best place (not sure... also we probably want to move the telemetry logic to the same place)?

**Comment by dmitryduev** [comment]
@kptkin thank you for a thorough review ( so typical though :) )!
I rm'ed raising errors and just left the `wandb.run is not None` checks to make `mypy` happy + added a check/raise error to the estimator hook's begin method, like you suggested.

**Review by kptkin — APPROVED** [review]
Awesome! Love the changes! 
You have the 🪄b-touch :)
